### ISSUE TITLE
Reintroduce TTL/expiration on documents

### ISSUE THREAD

**Issue by coffeemug** [rethinkdb/rethinkdb#746]
With secondary indexes it seems like it would be really easy to do. People create an timestamp sindex, and we periodically call delete on it.

**Comment by danielmewes** [comment]
@thelinuxlich suggested an interface for this in #3772:
> Something like r.table("foo").get(1).expires(seconds) would be super cool

**Comment by wojons** [comment]
I know i have made my own class that handles that stuff for me and will delete it on query that it does not exist but it cant delete others that are old with out a job i think something like could be done when we get to tables with max size because when the table reaches its max size there can be a trigger (yes i said it) that would fire off to try to clear space on the tabvle

**Comment by elifarley** [comment]
My suggestion is to initially implement a passive TTL, without any timers.

When a document is requested, its TTL is checked, and in case the object is too old, it is immediately deleted and the client receives a NOT FOUND error or something to that effect.

I have implemented this idea in (PassiveTimeOutMap)[https://issues.apache.org/jira/browse/COLLECTIONS-241] - here's a brief description:

>This is a Map decorator which passively evicts expired keys once their expiry time has been reached.

>When putting a key-value pair in the map, this decorator calls expiryTime(key, value), passing the key and the value as parameters, and uses the returned value as the expiry time for that key.

>When getting the value for a key, its expiry time is checked, and if it's greater than the current time, the value is returned. Otherwise, the key is removed from the decorated map, and null is returned.

>In doing so, there's no need to have a separate, active thread (hence the name *passive*) to check expiry times - the check is performed on demand.

**Comment by thelinuxlich** [comment]
This alone would solve many cases

**Comment by ravi** [comment]
The proposal above (from @elifarley) is great, but it fulfils one (but only one), the first, of three requirements I can think of:

1. Conceptually expire stale data.
2. Clear data in order to reduce disk space or improve performance.
3. React to expiration of data.

(I am thinking here of the features supported by Redis, and how I use them)

**Comment by thelinuxlich** [comment]
But at least for use cases like caching(redis.get(key) if null redis.set(key,value)) it can work

**Comment by ravi** [comment]
True.

**Comment by coffeemug** [comment]
Moving to subsequent, since there is a lot of demand. We'll see if we can get this scheduled soon.

**Comment by thelinuxlich** [comment]
Great!

**Comment by marshall007** [comment]
Once this lands, it would be pretty neat if you could subscribe to a changefeed specifically for TTL expirations (as opposed to all deletions on the table).

**Comment by coffeemug** [comment]
> it would be pretty neat if you could subscribe to a changefeed specifically for TTL expirations

:+1: that would be awesome!

**Comment by meznaric** [comment]
Looking forward to see this implemented!

**Comment by hsen-dev** [comment]
+1

**Comment by bithavoc** [comment]
:+1:

**Comment by codinfox** [comment]
👍 Looking forward to this feature!!

**Comment by freak4pc** [comment]
Is this still happening?

**Comment by coffeemug** [comment]
It isn't currently scheduled, but we do plan to incorporate it into the product in the future. Unfortunately no ETA yet.

**Comment by ravi** [comment]
This feature would make RethinkDB valuable for use cases like caching, messaging and service measurement and monitoring / time series data.

**Comment by coffeemug** [comment]
Definitely. It can already be simulated by creating an index on a time field, and periodically deleting old indexed entries. It's a little bit of a pain, so we're going to add some porcelain commands/features to make this easier, but it's definitely doable now.

**Comment by mmarmol** [comment]
This and a proper Java Driver would make rethinkdb the strongest candidate to replace our Couchbase Cluster. +1

**Comment by danielmewes** [comment]
@mmarmol We're currently working on the Java driver: https://github.com/rethinkdb/rethinkdb/issues/3930 . Would the TTL work-around with a secondary index on a timestamp and a regularly run cleanup job work for you?

**Comment by mmarmol** [comment]
If the regularly clean up is done by rethinkdb them yes, it would be fine. If I need to run the regularly clean them will not be that great, probably a problem. We use a lot TTL cause we have around 90% of our items created with that and we would like to have it seamless managed.

Cheers,

**Comment by thelinuxlich** [comment]
I agree with @mmarmol, the expiration should run on RethinkDB side to avoid another layer to take care of the risk of leaving documents on unsynced state

**Comment by ravi** [comment]
I too would prefer RethinkDB to handle this, primarily because it would likely be a lot more optimised, than an external process kicking in and removing data. And of course there is also the operational complexity.

**Comment by NicolasLM** [comment]
Having the 3 points presented by @ravi directly included in RethinkDB would save some work to users. Not having to implement a Celery or cron task to expire documents would be great.

**Comment by eav** [comment]
+1

**Comment by Dumbear** [comment]
+1

**Comment by jpillora** [comment]
+1

**Comment by thomasmodeneis** [comment]
+1

**Comment by stresslimit** [comment]
+1

**Comment by SlippingAway** [comment]
+1

**Comment by ardoi** [comment]
+1

**Comment by dpalmisano** [comment]
+1

**Comment by geoah** [comment]
+1 This would be very useful indeed.

**Comment by danielmewes** [comment]
I think we'll be able to tackle this relatively soon. RethinkDB 2.3 or 2.4ish most likely.

**Comment by webmasterkai** [comment]
## Request

Enable the "ttl garbage collector" task to run a user defined query periodically instead of having a fixed delete action. See #3636, #1863, and #3740 for background on user defined/serialized ReQL.

Being able to define queries more advanced than a fixed equivalent to `r.table('foo').filter(r.row('expireAfter').lt(r.now())).delete()` would be excellent and much more powerful.

### Use cases beyond a simple `ttl` field:
* Ability to look at any field instead of a specific ttl attribute.
* Instead of a full delete, update a field value.
* Auto delete users and _related data from multiple tables_ when a user hasn't logged in within the last year.
* Auto update table content from remote API using `r.http()`
* New query can retroactively remove "expired" rows and doesn't depend on a ttl flag on `tableCreate`.
* When the field `spamScore` reaches a certain threshold delete it. (changfeed reactions on the row living inside the db?)

Alternatively I'd just vote to have the ttl expire trigger an event (#3367?) so I can process it via `changes` however I like.

**Comment by mmarmol** [comment]
@webmasterkai brings an excellent point "Instead of a full delete, update a field value." Other databases offering ttl work like that. Ideally first a logical delete and when ever the database has a chance or time do an actual physical delete, ex. indexing or any other maintenance batch task on the table.

**Comment by coffeemug** [comment]
@webmasterkai -- you proposal is very cool, but a bit complicated by the fact a full table scan with a `filter` is a fairly inefficient way of doing this; for pure TTL we'd instead create an index on the expiration field instead.

I also like the idea of triggering a feed event on expiration -- it has a certain elegance to it (though would probably not be as efficient as running a batch ttl delete on the server).

Since this is a version or two away, we'll explore all these options during the ReQL discussion period when this gets scheduled. This is pretty exciting -- it's always fun to try and tease out an elegant, efficient solution when there are lots of options :grin:

**Comment by geoah** [comment]
I would agree with @webmasterkai on the additional uses and I would like to emphasise the changefeed reactions to expiries and times in general.

Something like `r.table('foo').changes(r.row('startsAt').eq(r.now()).run()` would really be useful in so many cases.
Or even with time calculations `r.table('foo').changes(r.row('startsAt').eq(r.now().addMinutes(5)).run()` to find anything that is starting in 5 minutes. (`addMinutes` is just a dummy method as I was too lazy to see if a real one exists in rethinkdb).

+1 for plugins btw~ :D

**Comment by mbroadst** [comment]
+1 I feel this + raft could easily allow me to use rethinkdb for service discovery (instead of e.g. etcd), which would be great

**Comment by devlounge** [comment]
Hi,

Any update on the TTL feature? (auto delete based on indexed timestamp field)

**Comment by coffeemug** [comment]
This has not been scheduled yet. We will get it done, but it will likely take a few more releases since there are more pressing issues, and this is easily simulated by creating an index and periodically running a delete query.

**Comment by larkost** [comment]
I talked to @danielmewes offline, and mentioned an algorithm that might work to keep the amount of work needed to handle this on the server:
1. Require that the comparison with the `now()`-esk function be made on an index. This will keep us from having to do table scans.
2. In the data structure backing the changefeed record the value after the one we last reported on.
3. Set a timer for when that time will come.
4. If a record is changed, then processes it immediately if it is before our current time, otherwise ignore it.
5. When the timer fires process all of the records in that index and send them out. It might be worth waiting some small time interval (at most 1 second) to coalesce a few data points together. Note that it is possible that the record we were waiting for was deleted, and if there are no records for this index point, this is almost a no-op.
6. Advance to the next point on the index and loop back to item 3.

I think that this proposal would require a separate command from `r.now()`, partially because it would require an index to process, and partially because it would not just be evaluated once.

Also, this concept is going to suffer the same clock skew issues as talked about in #5000.

**Comment by tatsujin1** [comment]
+1 "Instead of a full delete, update a field value."
Getting a changefeed event feels much more flexible. And also useful in my case :)

**Comment by ravi** [comment]
Updating fields instead of delete is fine if it is an additional feature, but IMHO it cannot supplant the basic need or use case for automatic document expiration. The idea and power (again IMHO) of the latter lies in the fact that an important part of certain usage patterns can be entrusted to the server to be performed in an optimal way. For it to be truly useful, it should not require programmatic action outside of initial TTL configuration.

**Comment by mbroadst** [comment]
+1 to that

**Comment by coffeemug** [comment]
:+1: agree with @ravi. Feeds on ttl are great to notify the clients when a document is expired, but I think the feature wouldn't be very useful unless the db automatically handles deletion of expired documents.

**Comment by tatsujin1** [comment]
Certainly, if handing expiration is the goal - meaning the action is to delete - it's probably more efficient to do that directly on the server. However, I imagine this feature could be a generic trigger for "time to act", whatever that may be. Which in my case would very handy.

**Comment by tatsujin1** [comment]
Maybe the api (somewhere) could specify an argument `action='delete'` or `action={ "field": "value"}` (python syntax), that would be awesome. Default action might be `delete`.

**Comment by marshall007** [comment]
:+1: to what @tatsujin1 said, I was going to mention the same use-case. We store generic tasks in the db and currently poll to see which ones need to be run.

**Comment by saary** [comment]
@tatsujin1 +! to setting the actions
Another useful action can be: `{ 'http': { ...} }` this allows to send out http request directly to a cluster and have **one** of the servers react to them

**Comment by coffeemug** [comment]
@danielmewes -- this would be super-useful for Fusion. Would it be possible to include it into 2.3?

**Comment by mlucy** [comment]
@coffeemug -- this particular feature would probably be a *lot* easier to implement in Fusion itself.  (Although @danielmewes probably has a better sense of how difficult it is on the server than I do.)  You can give every row a  timestamp field (which might be useful anyway), create a secondary index on it, and periodically call `delete` on the oldest ones.

**Comment by coffeemug** [comment]
Talked to @danielmewes and @mlucy -- we agreed we'll skip this for 2.3.

**Comment by ravi** [comment]
Oh no :(. The wait continues! :)

And to make this comment productive: I want to reiterate that IMVHO this feature will be most useful if it is embedded as a core function in the server, where the code can make use of internal timers, and other optimisations and considerations, so that it will exact the least performance penalty.

**Comment by thelinuxlich** [comment]
Not only performance, but guarantees that it will remove the expired row

**Comment by ravi** [comment]
@thelinuxlich indeed, that's an important point.

**Comment by spikying** [comment]
+1 for TTL feature.

**Comment by eduardonunesp** [comment]
+ 1 TTL :)

**Comment by let4be** [comment]
Would be extremely useful for my application
thing is software that is responsible for manual TTL deletes may go temporary down(or the whole machine) and in this case TTL will stuck in database for a while., depending on use case this may cause problems

**Comment by LeDominik** [comment]
:+1: for TTL -- doing it manually is painful and prone to error because in general applications have any kind of batch-layer more as a last thought :smile: and you just have to forget your filter in one query and it get's weird. Maybe contrary to #4008 I would keep it real simple and just store the timestamp-of-death on the object itself. Makes for very simple filtering / scanning logic and the API can surface the complex stuff. 

E.g. [HBase](https://hbase.apache.org/book.html#ttl) does it even hierarchical on column-family as well as cell-level (but they switched to msec notation in the newer cell-level so this seems to be the current way of thinking).

**Comment by xenomuta** [comment]
+1 for TTL :+1:

**Comment by bhurlow** [comment]
:+1:

**Comment by bardu** [comment]
TTL would be awesome

**Comment by tatsujin1** [comment]
Maybe not the best place, but I added another suggestion for API to #4008.

**Comment by StephanMeijer** [comment]
+1 for an expire-feature.

**Comment by sinfex** [comment]
+1

**Comment by chovy** [comment]
any eta on ttl? its been 3 years.

**Comment by danielmewes** [comment]
@chovy No definite ETA yet, but this is definitely on our short-list of things to implement soon. I'm pretty sure we will be able to ship this this year.

**Comment by chovy** [comment]
@danielmewes thanks for the update. Looking forward to it. I'll add my use case to the list:

each row in the table would have its own expiring field set. I have a list of messages are going to expire based on what the user has defined the expiration in their profile settings -- could be "delete all messages after 10 minutes" or an option when creating a message "delete this message after [5] minutes". I read that some proposed solutions are setting the TTL on the table, that won't work for my usecase, i need tuple level control.

**Comment by StephanMeijer** [comment]
> No definite ETA yet, but this is definitely on our short-list of things to implement soon. I'm pretty sure we will be able to ship this this year.

Great news! Thanks.

**Comment by tatsujin1** [comment]
"And there was much rejoicing" 😁

**Comment by devlounge** [comment]
So what you and most of us would like is a background ttl task which deletes or marks for deletion and deletes multiple times a day, based on a secondary index whenever its value(s) match a user defined expression which evaluates to True. 

Anthony

**Comment by StephanMeijer** [comment]
Or something like a trigger:

when `current_time` is after `timestamp of expirement of row`: delete the row.

Or maybe something like making the specific row inaccessible after the timestamp is passed, and deleting it once in a few hours.

**Comment by thelinuxlich** [comment]
Once we get reliable changefeeds, this could be implemented on a system table without much difficult

**Comment by thomasmodeneis** [comment]
For sure if there is reliable changefeeds this feature will be easily implemented :)

**Comment by danielmewes** [comment]
I'm actually not sure how useful reliable / resumable changefeeds will be for this.
If you have an index on the timestamp, re-running the query periodically or when restarting is cheap.

For using changefeeds, an important missing feature is having a changefeed on `r.now`. https://github.com/rethinkdb/rethinkdb/issues/3583

**Comment by danielmewes** [comment]
A note if you're reading this and want to implement TTL manually until it's available natively in RethinkDB:
  There's a description of how to do it in this thread on Google Groups: https://groups.google.com/forum/#!topic/rethinkdb/tFSiG5Ex1KE

**Comment by StephanMeijer** [comment]
Thank you, Daniel.

**Comment by jpillora** [comment]
In an effort to reduce spam, I think this thread should be locked and users in favour should leave a :+1: reaction on coffeemug's [issue description comment](https://github.com/rethinkdb/rethinkdb/issues/746#issue-13880815). (Sorry for the spam!)

**Comment by danielmewes** [comment]
Locking the issue will stop us from getting useful extra information. But using the :+1: on the initial comment instead of posting separate messages would be great!

Please use the :+1: reaction on the topmost comment to express your support for this feature
------

**Comment by danielmewes** [comment]
We're now working on a concrete proposal for this feature and should have something up here for general discussion within the next days.

**Comment by danielmewes** [comment]
The proposals are up:
https://github.com/rethinkdb/rethinkdb/issues/5823 is the core one, and https://github.com/rethinkdb/rethinkdb/issues/5813 supplements it.

**Comment by chovy** [comment]
any news on this? i assume it hasn't landed yet

**Comment by danielmewes** [comment]
@chovy It's landing in the upcoming major release 2.4.0. See the two issues linked in my previous comment.

**Comment by danielmewes** [comment]
Closing this in favor of https://github.com/rethinkdb/rethinkdb/issues/5823 (together with https://github.com/rethinkdb/rethinkdb/issues/5813 ) which describes the actual API for this functionality.
Will ship with 2.4.0.

**Comment by ultimatecoder** [comment]
Is 2.4 released? I am in need of this feature. Thanks!
